Doc/tools/extensions/refwarn.py

```diff
@@ -0,0 +1,58 @@
+"""Capture and report missing cross-references in the documentation."""
+
+import csv
+import os.path
+from collections import Counter
+
+
+def _rpr(obj, /):
+    """Return the repr without the enclosing quotation marks."""
+    return repr(obj)[1:-1]
+
+
+def get_node_source(node):
+    """Return the source file for 'node'."""
+    while node:
+        if node.source:
+            return node.source
+        node = node.parent
+    return "<unknown>"
+
+
+def warn_missing_reference(app, domain, node):
+    warnfile = os.path.join(app.outdir, 'refwarn.csv')
+
+    typ = _rpr(node['reftype'])
+    target = _rpr(node['reftarget'])
+    domain = _rpr(node.get('refdomain', 'std'))
```

Comment:
`type_` is a often used:

```suggestion
    type_ = _repr(node['reftype'])
    target = _repr(node['reftarget'])
    domain = _repr(node.get('refdomain', 'std'))
```